\documentclass[a4paper,UKenglish]{lipics-v2016}

\usepackage{microtype}

\bibliographystyle{plainurl}

% Author macros::begin %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\title{Indexing Variation Graphs\footnote{This work was supported by the Wellcome Trust grant [098051].}}

\author[1]{Jouni Sirén}
\affil[1]{Wellcome Trust Sanger Institute, Hinxton, Cambridge, UK\\
  \texttt{jouni.siren@iki.fi}}
\authorrunning{J. Sirén}

\Copyright{Jouni Sirén}

\subjclass{E.1 Data Structures}
\keywords{Burrows-Wheeler transform, de Bruijn graphs, path indexes, variation graphs}
% Author macros::end %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Editor-only macros:: begin (do not touch as author)%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\EventEditors{John Q. Open and Joan R. Acces}
\EventNoEds{2}
\EventLongTitle{42nd Conference on Very Important Topics (CVIT 2016)}
\EventShortTitle{CVIT 2016}
\EventAcronym{CVIT}
\EventYear{2016}
\EventDate{December 24--27, 2016}
\EventLocation{Little Whinging, United Kingdom}
\EventLogo{}
\SeriesVolume{42}
\ArticleNo{23}
% Editor-only macros::end %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% Mathematics
\newcommand{\set}[1]{\ensuremath{\{ #1 \}}}
\newcommand{\abs}[1]{\ensuremath{\lvert #1 \rvert}}
\newcommand{\Oh}[1]{\ensuremath{\mathsf{O}\!\left( #1 \right)}}

% DNA
\newcommand{\dnaseq}[1]{\ensuremath{\mathtt{#1}}}
\newcommand{\baseA}{\dnaseq{A}}
\newcommand{\baseC}{\dnaseq{C}}
\newcommand{\baseG}{\dnaseq{G}}
\newcommand{\baseT}{\dnaseq{T}}
\newcommand{\baseN}{\dnaseq{N}}
\newcommand{\dnacomp}[1]{\ensuremath{\overline{#1}}}
\newcommand{\revcomp}[1]{\ensuremath{\overleftarrow{#1}}}

% Queries
\newcommand{\rank}{\ensuremath{\mathsf{rank}}}
\newcommand{\select}{\ensuremath{\mathsf{select}}}
\newcommand{\LF}{\ensuremath{\mathsf{LF}}}

% Graphs
\newcommand{\gindegree}{\ensuremath{\mathsf{in}}}
\newcommand{\goutdegree}{\ensuremath{\mathsf{out}}}
\newcommand{\glabel}{\ensuremath{\mathsf{label}}}
\newcommand{\gkey}{\ensuremath{\mathsf{key}}}
\newcommand{\gvalue}{\ensuremath{\mathsf{value}}}
\newcommand{\gnode}{\ensuremath{\mathsf{node}}}

% Shorthands
\newcommand{\kmer}[1]{$#1$\nobreakdash-mer}
\newcommand{\kcollection}[1]{$#1$\nobreakdash-collection}
\newcommand{\orderk}[1]{order\nobreakdash-$#1$}
\newcommand{\LFmapping}{LF\nobreakdash-mapping}
\newcommand{\FMindex}{FM\nobreakdash-index}

% Structures
\newcommand{\SA}{\ensuremath{\mathsf{SA}}}
\newcommand{\BWT}{\ensuremath{\mathsf{BWT}}}
\newcommand{\Carray}{\ensuremath{\mathsf{C}}}
\newcommand{\LCP}{\ensuremath{\mathsf{LCP}}}
\newcommand{\bvIN}{\ensuremath{\mathsf{IN}}}
\newcommand{\bvOUT}{\ensuremath{\mathsf{OUT}}}


\begin{document}

\maketitle

\begin{abstract}
Variation graphs, which represent genetic variation within a population, are replacing sequences as reference genomes. Path indexes are one of the most important tools for working with variation graphs. They generalize text indexes for graphs, allowing one to find the paths matching the query string. We propose using pruned de Bruijn graphs as path indexes, encoding them space-efficiently with the Burrows-Wheeler transform. We also generalize many ideas from text indexing literature to work with graphs. The proposed approach has been implemented in the variation graph toolkit vg.
\end{abstract}


\section{Introduction}

1000GP: \cite{1000GP2015}

Graph genomes: \cite{Schneeberger2009}

FM-index: \cite{Ferragina2005a}

CST-NPR: \cite{Fischer2009a}

CST MEMs: \cite{Ohlebusch2010a}

XBW: \cite{Ferragina2009b}

GCSA: \cite{Siren2014}

Succinct DBG: \cite{Bowe2012}, variable-order \cite{Boucher2014}, alternative representation \cite{Roedland2013}

BWBBLE: \cite{Huang2013}

Contracted DBG: \cite{Cazaux2014}

Compressed DBG: \cite{Marcus2014}

Hypertext index: \cite{Thachuk2013}

FM-index of alignment: \cite{Na2015}

HISAT / HISAT2: \cite{Kim2015}


\section{Background}

\subsection{Strings}\label{sect:strings}

A \emph{string} $S[0, n-1] = s_{0} \dotsm s_{n-1}$ of length $\abs{S} = n$ is a sequence of \emph{characters} over an \emph{alphabet} $\Sigma = \set{0, \dotsc, \sigma - 1}$. For indexing purposes, we often consider \emph{text} strings $T[0, n-1]$ terminated by an \emph{endmarker} $T[n-1] = \$ = 0$ not found anywhere else in the text. \emph{Binary} sequences are strings over the alphabet $\set{0, 1}$. A \emph{substring} of string $S$ is a sequence of the form $S[i, j] = s_{i} \dotsm s_{j}$. We call substrings of the type $S[0, j]$ and $S[i, n-1]$ \emph{prefixes} and \emph{suffixes}, respectively, and substrings of length $k$ as \kmer{k}s. We say that string $S'$ is a substring of string collection $\mathcal{S}$, if there is a string $S \in \mathcal{S}$ such that $S'$ is a substring of $S$.

In some cases, we also consider potentially \emph{infinite} character sequences $S = (s_{i})_{i \in Z}$, where set $Z$ is a contiguous infinite subset of $\mathbb{Z}$. The notion of substrings generalizes for such infinite sequences in a natural way. We say that a substring of an infinite sequence $S$ is \emph{left-infinite} if it extends infinitely to the left, and \emph{right-infinite} if it extends infinitely to the right. A substring of a finite or infinite sequence $S$ is \emph{left-maximal} if it is left-infinite or a prefix; \emph{right-maximal} if it is right-infinite or a suffix; and \emph{maximal} if it is both left-maximal and right-maximal.

In this paper, we primarily consider sequences over the \emph{DNA} alphabet $\set{\$, \baseA, \baseC, \baseG, \baseT, \baseN}$. Characters $\baseA$, $\baseC$, $\baseG$, and $\baseT$ are called \emph{bases}, while character $\baseN$ represents an arbitrary or unknown base. In addition to the endmarker $\$$, the alphabet may also contain other characters for technical purposes. Each character $c$ of the DNA alphabet has a \emph{complement} $\dnacomp{c}$ defined as $\dnacomp{\baseA} = \baseT$, $\dnacomp{\baseC} = \baseG$, $\dnacomp{\baseG} = \baseC$, $\dnacomp{\baseT} = \baseA$, and $\dnacomp{c} = c$ for other characters $c$. Given a DNA sequence $S$, its \emph{reverse complement} is the sequence $\revcomp{S}$ obtained by reversing the non-technical parts of the sequence and replacing each character with its complement. For example, $\revcomp{\dnaseq{GATTACA}\$} = \dnaseq{TGTAATC}\$$.

Given string $S[0, n-1]$, we define $S.\rank(i, c)$ to be the number of occurrences of character $c$ in the prefix $S[0, i-1]$. We also define $S.\select(i, c) = \arg \max_{j \le n} S.\rank(j, c) < i$ as the position of the occurrence of character $c$ with rank $i > 0$.\footnote{These definitions correspond to the ones used in the SDSL library \cite{Gog2014b}.} A \emph{bitvector} is a binary sequence supporting efficient $\rank$/$\select$ queries. \emph{Wavelet trees} \cite{Grossi2003} are space-efficient data structures that use bitvectors to support $\rank$/$\select$ queries on arbitrary strings.

The following definitions are useful when discussing text indexes and path indexes.

\begin{definition}[Prefix-matching strings]
Let $S$ and $S'$ be strings over alphabet $\Sigma$. We say that string $S$ and $S'$ \emph{prefix-match}, if $S$ is a prefix of $S'$ or $S'$ is a prefix of $S$.
\end{definition}

\begin{definition}[Prefix-free set]
Let $\mathcal{S}$ be a set of strings over alphabet $\Sigma$. We say that set $\mathcal{S}$ is \emph{prefix-free}, if no two strings $S, S' \in \mathcal{S}$ (with $S \ne S'$) prefix-match.
\end{definition}

\subsection{Text indexes}

The \emph{suffix tree} \cite{Weiner1973} is the most fundamental full-text index supporting arbitrary substring queries. It is formed by taking the suffixes of the text, storing them in a trie, and compacting unary paths in the trie into single edges. Although fast and versatile, suffix trees are only of limited use in indexing large texts, as they require much more space than the text itself.

% FIXME SA/BWT/LCP example
\emph{Suffix arrays} \cite{Manber1993} were introduced as a space-efficient alternative to the suffix tree. The suffix array of text $T[0, n-1]$ is an array of pointers $\SA[0, n-1]$ to the suffixes of the text in \emph{lexicographic order}. Given a text and its suffix array, we can find the occurrences of \emph{pattern} string $X$ in the text by using \emph{binary search} in $\Oh{\abs{X} \log n}$ time. The suffix array requires $n \log n$ bits of space in addition to the $n \log \sigma$ bits used by the text --- still much more than the text --- while its functionality is more limited than that of the suffix tree.

The \emph{Burrows-Wheeler transform (BWT)} \cite{Burrows1994} is a permutation of the text with the same combinatorial structure as the suffix array. Given text $T[0, n-1]$, its Burrows-Wheeler transform is a string $\BWT[0, n-1]$, where $\BWT[i] = T[(\SA[i]-1) \bmod n]$. Given the \emph{lexicographic rank} $i$ of suffix $T[\SA[i], n-1]$, we can use \emph{\LFmapping} on the BWT to find the lexicographic rank of the previous suffix $T[(\SA[i]-1) \bmod n, n-1]$. Let
$$
\LF(i) = \Carray[\BWT[i]] + \BWT.\rank(i, \BWT[i]),
$$
where $\Carray[c]$ is the number of suffixes of the text starting with any character $c' < c$. Then $\SA[\LF(i)] = (\SA[i]-1) \bmod n$. We can also generalize the definition for any character $c \in \Sigma$:
$$
\LF(i, c) = \Carray[c] + \BWT.\rank(i, c).
$$
Let $X$ be a string. If there are $i$ suffixes $S'$ of text $T$ such that $S' < X$ in lexicographic order, then there are $\LF(i, c)$ suffixes smaller than string $cX$ in lexicographic order.

Given the similarities to the suffix array, we can use the BWT as a space-efficient text index. The \emph{\FMindex} \cite{Ferragina2005a} combines a plain or compressed representation of the BWT supporting rank/select queries, the $\Carray$ array, and a set of \emph{sampled pointers} from the suffix array. It finds the \emph{lexicographic range} of suffixes matching pattern $X$ (having $X$ as a prefix) with a process called \emph{backward searching}. If the lexicographic range matching suffix $X[i+1, \abs{X}-1]$ of the pattern is $\SA[sp, ep]$, then the lexicographic range matching suffix $X[i, \abs{X}-1]$ of the pattern is $\SA[\LF(sp, X[i]), \LF(ep+1, X[i]) - 1]$. Finding the lexicographic range matching the entire pattern requires $\Oh{\abs{X}}$ rank queries.

We can use the sampled suffix array pointers to find the text positions containing the occurrences. If $\SA[i]$ is not sampled, we start iterating $\LF(i)$, until we find a sampled pointer. If we find a sample at $\SA[\LF^{k}(i)]$, we know that
$$
\SA[i] = (\SA[\LF^{k}(i)] + k) \bmod n.
$$
If we have sampled one out of $d$ suffix array pointers at regular intervals, finding each occurrence takes $\Oh{d}$ rank queries. If we also sample one out of $d'$ \emph{inverse suffix array} pointers\footnote{The suffix array is a permutation of $\set{0, \dotsc, n-1}$, and the inverse suffix array is the inverse permutation.}, we can \emph{extract} an arbitrary substring $X$ of the text using $\Oh{\abs{X}+d'}$ rank queries. In a typical case, we expect an \FMindex{} to take less than $n \log \sigma$ bits of space and to be able to find about 100,000 pattern occurrences per second \cite{Ferragina2009a}.

The \emph{longest-common-prefix array} (LCP array) \cite{Manber1993} is an integer array $\LCP[0, n-1]$, where each value $\LCP[i]$ tells the length of the longest common prefix of suffixes $T[\SA[i-1], n-1]$ and $T[\SA[i], n-1]$ (with $\LCP[0] = 0$). It is often used to augment the functionality of the suffix array and the \FMindex. In particular, if we augment the \FMindex{} with the LCP array and the topology of the suffix tree, we get the \emph{compressed suffix tree}, which supports the full functionality of the suffix tree in a space-efficient manner \cite{Sadakane2007}.

\subsection{Graphs}\label{sect:graphs}

A \emph{graph} $G = (V, E)$ consists of a set of \emph{nodes} $V = \set{0, \dotsc, \abs{V}-1}$ and a set of \emph{edges} $E \subseteq V \times V$. We say that $(u, v) \in E$ is an edge \emph{from} node $u$ \emph{to} node $v$, and assume that the edges are \emph{directed}: $(u, v) \ne (v, u)$ for $u \ne v$. The \emph{indegree} $G.\gindegree(v)$ of node $v$ is the number of \emph{incoming} edges to $v$, while the \emph{outdegree} $G.\goutdegree(v)$ is the number of \emph{outgoing} edges from $v$.

The graphs we consider are \emph{labeled} with alphabet $\Sigma$: each node $v \in V$ has a \emph{label} $G.\glabel(v) \in \Sigma$. A \emph{path} in a graph is a sequence of nodes $P = v_{0} \dotsm v_{\abs{P}-1}$ such that $(v_{i}, v_{i+1}) \in E$ for all $i, i+1 \in \set{0, \dotsc, \abs{P}-1}$. We say that $v_{0}$ is the \emph{start} node and $v_{\abs{P}-1}$ is the \emph{end} node of the path. The label of a path is the concatenation of node labels $G.\glabel(P) = G.\glabel(v_{0}) \dotsm G.\glabel(v_{\abs{P}-1})$.

We generalize the definition for infinite paths $P = (v_{i})_{i \in Z}$ in a similar way as we did with infinite character sequences in Section~\ref{sect:strings}. We say that path $P$ is \emph{left-maximal} if it starts at the source node or extends infinitely to the left; \emph{right-maximal} if it ends at the sink node or extends infinitely to the right; and \emph{maximal} if it is both left-maximal and right-maximal.

We assume for convenience that all graphs have two special nodes: the \emph{source} node $s$ and the \emph{sink} node $t$. To distinguish them from the other nodes, we label them with characters $G.\glabel(s) = \#$ and $G.\glabel(t) = \$$. The label of the sink node is unique in the graph, while the label of the source node can be used in other nodes for technical purposes. We add an edge $(s, v)$ for all nodes $v \in V \setminus \set{s}$ that have no other incoming edges, and an edge $(v, t)$ for all nodes $v \in V \setminus \set{t}$ with no other outgoing edges. We also add the edge $(t, s)$ to guarantee that $G.\gindegree(v) \ge 1$ and $G.\goutdegree(v) \ge 1$ for all nodes $v \in V$. However, we assume for convenience that no path in the graph crosses the edge $(t, s)$.

In this paper, we work with de~Bruijn graphs and their generalizations. In order to do so, we need to define collections of (finite or infinite) sequences suitable for constructing \orderk{k} de~Bruijn graphs.

\begin{definition}[\kcollection{k}]
Let $\mathcal{S}$ be a collection of character sequences over alphabet $\Sigma$, and let $k$ be a parameter value. We say that $\mathcal{S}$ is a \emph{\kcollection{k}}, if the collection contains substrings $\#^{k}$ and $\$^{k}$, and each sequence $S \in \mathcal{S}$
\begin{enumerate}
\item is left-infinite or begins with $\#^{k}$;
\item is right-infinite or ends with $\$^{k}$; and
\item contains no other occurrences of characters $\#$ and $\$$.
\end{enumerate}
\end{definition}

% FIXME example of strings, DBG, long paths, false positives
\begin{definition}[de~Bruijn graph]
Let $\mathcal{S}$ be a \kcollection{k} over alphabet $\Sigma$. The \orderk{k} \emph{de~Bruijn graph} of $\mathcal{S}$ is a graph $G = (V, E)$ such that
\begin{enumerate}
\item each node $v_{X} \in V$ represent a distinct \kmer{k} $X$ occurring in $\mathcal{S}$, with $G.\glabel(v_{X}) = X[0]$;
\item each node $v_{X} \in V$ has a \emph{key} $G.\gkey(v_{X}) = X$; and
\item each edge $(v_{X}, v_{Y}) \in E$ represents a \kmer{k+1} $X[0]Y = Xc$ (where $c \in \Sigma$) occurring in $\mathcal{S}$.
\end{enumerate}
We use the node corresponding to $\#^{k}$ as the source node $s$ and the node corresponding to $\$^{k}$ as the sink node $t$, adding the technical edge $(t, s)$ in the usual way.
\end{definition}

The following lemmas summarize the key properties of de~Bruijn graphs for indexing purposes.

\begin{lemma}[Keys prefix-match with path labels]\label{lemma:dbg-key}
Let $G = (V, E)$ be an \orderk{k} de~Bruijn graph, and let $P$ be a path starting from node $v_{0} \in V$. Then strings $G.\gkey(v_{0})$ and $G.\glabel(P)$ prefix-match.
\end{lemma}

\begin{proof}
Let $P[0, k'-1] = v_{0} \dotsm v_{k'-1}$ be a prefix of path $P$, where $1 \le k' \le \min(k, \abs{P})$.
If $k' = 1$, we know that $G.\glabel(P[0]) = G.\glabel(v_{0})$ is a prefix of $G.\gkey(v_{0})$.

Now assume that $k' > 1$. By induction, we know that $G.\glabel(P[1, k'-1])$ is a prefix of $G.\gkey(v_{1})$. As edge $(v_{i}, v_{i+1}) \in E$ exists, $G.\glabel(v_{0}) \cdot G.\gkey(v_{1}) = G.\gkey(v_{0}) \cdot c$ (where $c \in \Sigma$), and hence $G.\glabel(P[0, k'-1])$ is a prefix of $G.\gkey(v_{0})$. As the first $\min(k, \abs{P})$ characters of $G.\gkey(v_{0})$ and $G.\glabel(P)$ are identical, the result follows.
\end{proof}

\begin{lemma}[No false negatives]\label{lemma:dbg-fn}
Let $G = (V, E)$ be the \orderk{k} de~Bruijn graph of \kcollection{k} $\mathcal{S}$, and let $X$ be a finite substring of $\mathcal{S}$ containing at most one $\$$. Then there is a path $P$ in the graph with $G.\glabel(P) = X$.
\end{lemma}

\begin{proof}
Let $X = S[i, i+\abs{X}-1]$ for a sequence $S \in \mathcal{S}$. For $i \le j \le i+\abs{X}-1$, there is a node $v_{j} \in V$ representing substring $S[j, j+k-1]$, with $G.\glabel(v_{j}) = S[j]$. By definition, the de~Bruijn graph has an edge $(v_{j}, v_{j+1})$ for $i \le j < i+\abs{X}-1$. Hence $P = v_{i} \dotsm v_{i+\abs{X}-1}$ is a path in $G$ and $G.\glabel(P) = X$.
\end{proof}

\begin{lemma}[No false positives with $\abs{P} \le k$]\label{lemma:dbg-fp}
Let $G = (V, E)$ be an \orderk{k} de~Bruijn graph of \kcollection{k} $\mathcal{S}$, and let $P = v_{0} \dotsm v_{\abs{P}-1}$ be a path in $G$ with $\abs{P} \le k$. Then the label of the path is a prefix of $G.\gkey(v_{0})$ and hence a substring of a sequence $S \in \mathcal{S}$.
\end{lemma}

\begin{proof}
The label $G.\glabel(P)$ prefix-matches $G.\gkey(v_{0})$ by Lemma~\ref{lemma:dbg-key}. As $\abs{P} \le k$, string $G.\glabel(P)$ is a prefix of $G.\gkey(v_{0})$
\end{proof}

Note that the labels of all paths of length $k+1$ are also substrings of collection $\mathcal{S}$.

\subsection{Generalized indexes}

Suffix trees, suffix arrays, and \FMindex{} can be generalized to index multiple texts. There are also generalizations for other combinatorial structures. The \emph{XBW transform} \cite{Ferragina2009b} is essentially an \FMindex{} for \emph{labeled trees}. The nodes of the tree are sorted by path labels from the node to the root of the tree. $\BWT$ stores the labels of the children of each node (leaf nodes require special treatment). If a node has $k$ children, we encode that as binary sequence $0^{k-1} 1$. We concatenate these sequences to form bitvector $B$, which is used for mapping lexicographic ranks of nodes to the corresponding BWT ranges. The labels of the children of node $i$ are found in $\BWT[B.\select(i, 1) + 1, B.\select(i + 1, 1)]$ (with $0$ as the lower bound for $i = 0$).

% FIXME LF(i, c) picture from the slides for the earlier DBG example
The \emph{generalized compressed suffix array} (GCSA) \cite{Siren2014} is a further generalization of the XBW transform for a class of graphs that includes \emph{directed acyclic graphs} and de Bruijn graphs. Before a graph can be indexed, we have to transform it into an equivalent graph, where the nodes can be unambiguously sorted by the labels of the paths starting from them. The transformation is expensive and may increase the size of the graph exponentially. After sorting the nodes of the transformed graph, we encode them using three sequences. $\BWT$ contains the labels of the predecessors, while bitvectors $\bvIN$ and $\bvOUT$ encode the indegrees and outdegrees of each node in the same way as bitvector $B$ of the XBW transform. \LFmapping{} uses select queries on bitvector $\bvIN$ to map nodes into BWT ranges, ordinary \LFmapping{} with $\BWT$ to map incoming edges into the corresponding outgoing edges, and rank queries on bitvector $\bvOUT$ to map the outgoing edges to the predecessor nodes.


\section{Path indexes}

A \emph{path index} is a generalization of text indexes for \emph{labeled graphs}. Given a path index for graph $G = (V, E)$, we want to use the index to find the start nodes $v_{0} \in V$ of the paths $P = v_{0} \dotsm v_{\abs{X}-1}$ \emph{matching} the pattern $X$ (paths $P$ with $G.\glabel(P) = X$).

The design of a path index is a trade-off between the maximum supported query length and the size of the index. In the worst case, a graph may have up to $\sigma^{k}$ distinct labels for paths of length $k$ (ignoring the special treatment of characters $\#$ and $\$$), and the length of possible paths is unlimited in cyclic graphs. GCSA \cite{Siren2014} tried to avoid this by considering only a relatively simple class of graphs, using \emph{pruning heuristics} to simplify the graphs when necessary. The price of this was the possibility of \emph{false negatives}: path labels that exist in the graph but not in the index.

\subsection{Basic indexes}

The \emph{\kmer{k} index} is the simplest path index. In its most basic form, the \kmer{k} index consists of a parameter value $k$, a hash function $h$, and an array $A$ of \emph{key-value pairs} $(X, V_{X})$, where $X \in \Sigma^{k}$ and $V_{X} \subseteq V$. Given a \kmer{k} $X \in \Sigma^{k}$, we start looking for key $X$ from $A[h(X)]$, and list the set of start nodes $V_{X}$ as matches if we find it. Searching using a hash table is fast, but we can only search for patterns of length $k$, and the array requires a lot of space.

Another representation of the \kmer{k} index trades query performance for the ability to search for shorter patterns. Instead of using a hash table, we sort the key-value pairs by their keys, and store the pairs in array $A$ in sorted order. Given a pattern $X \in \Sigma^{\ast}$, $\abs{X} \le k$, we use binary search to find the range $A[sp, ep]$ of pairs $(X_{i}, V_{X_{i}})$, where pattern $X$ is a prefix of the key $X_{i}$. We then list the union $\bigcup_{i=sp}^{ep} V_{X_{i}}$ as the set of matches.

As de~Bruijn graphs and \kmer{k} indexes are both based on \kmer{k}s, we can use one to simulate the other. In order to define the de~Bruijn graph of a graph, we build a \kcollection{k} based on the collection $\mathcal{S}$ of the labels of the maximal paths in graph $G = (V, E)$. If sequence $S \in \mathcal{S}$ is the label of path $P = (v_{i})_{i \in Z}$ in the graph, we set $\mathcal{S}.\gnode(S, i) = v_{i}$ for all positions $i \in Z$. We then transform $\mathcal{S}$ into a \kcollection{k} by adjusting the number of characters $\#$ at the beginning of each non-left-infinite sequence, and the number of characters $\$$ at the end of each non-right-infinite sequence. If $S[i]$ is a $\#$ we inserted, we set $\mathcal{S}.\gnode(S, i) = (s, j)$, where $s$ is the start node of $G$ and $j$ is the distance to the nearest non-inserted $\#$ in $S$.

\begin{definition}[de~Bruijn graph of a graph]
Let $G = (V, E)$ be a labeled graph, and let $\mathcal{S}$ be the \kcollection{k} of maximal path labels in $G$. Then the \orderk{k} de~Bruijn graph of $\mathcal{S}$ is the \orderk{k} de~Bruijn graph of graph $G$.
\end{definition}

Given a \kmer{k+1} index of graph $G = (V, E)$, we can use it to simulate the \orderk{k} de~Bruijn graph $G' = (V', E')$ of $G$. As the nodes of the de~Bruijn graph correspond to \kmer{k}s and the edges correspond to \kmer{k+1}s, we can represent the nodes by the adjacent edges. If we are interested in node $v_{X} \in V'$, we search for incoming edges $cX$ and outgoing edges $Xc$ for all $c \in \Sigma$. Determining the existence of a node requires $2 \sigma$ queries in the \kmer{k+1} index, and we learn the presence of adjacent nodes and edges in the process.

On the other hand, we can use the \orderk{k} de~Bruijn graph $G' = (V', E')$ of graph $G = (V, E)$ as a \kmer{k} index of $G$ that supports queries of arbitrary length. For each key-value pair $(X, V_{X})$ in the index, we find the node $v_{X} \in V'$ with $G'.\gkey(v_{X}) = X$, and attach the \emph{value} $G'.\gvalue(v_{X}) = V_{X}$ to the node. We say that a pattern matches node $v \in V'$ if it prefix-matches string $G.\gkey(v)$. As Lemmas~\ref{lemma:dbg-fn} and \ref{lemma:dbg-fp} suggest, there will be no false negatives, and no \emph{false positives} (path labels that exist in the de~Bruijn graph but not in the input graph) with patterns of length at most $k$. With longer patterns, we have to \emph{verify} the results in graph $G$ to avoid false positives. We will develop the idea of using graphs similar to de~Bruijn graphs as path indexes further in the next section.

\subsection{Path graphs}

De~Bruijn graphs are a special case of path graphs. They represent the paths of length $k$ in the \emph{input graph} as nodes and the pairs of paths overlapping on $k-1$ nodes of the input graph as edges. We can generalize the notion by allowing the nodes of the path graph to represent paths of different lengths.

\begin{definition}[Path graph]
Let $\mathcal{S}$ be a \kcollection{k} of the labels of maximal paths in graph $G = (V, E)$, and let $\mathcal{K}$ be a prefix-free set of strings containing $\#^{k}$ and $\$^{k}$. We assume that each right-maximal substring $S$ in $\mathcal{S}$ prefix-matches a string $K \in \mathcal{K}$, and that each string $K \in \mathcal{K}$ is a substring of $\mathcal{S}$. Then the \orderk{k} \emph{path graph} of graph $G$ with \emph{key set} $\mathcal{K}$ is a graph $G' = (V', E')$ such that
\begin{enumerate}
\item each node $v_{K} \in V'$ represent a distinct key $K \in \mathcal{K}$, with $G'.\glabel(v_{K}) = K[0]$;
\item each node has a key $G'.\gkey(v_{K}) = K$ and a value $G'.\gvalue(v_{K}) = V_{K}$, where $V_{K}$ is the set of nodes $\mathcal{S}.\gnode(S, i)$ for $S \in \mathcal{S}$ and positions $i$ such that $S[i, i+\abs{K}-1] = K$; and
\item each edge $(v_{K}, v_{K'}) \in E'$ represents the occurrence of substring $K[0] K'$ in $\mathcal{S}$ such that strings $K$ and $K[0] K'$ prefix-match.
\end{enumerate}
We use the node corresponding to $\#^{k}$ as the source node $s$ and the node corresponding to $\$^{k}$ as the sink node $t$, adding the technical edge $(t, s)$ in the usual way.
\end{definition}

\begin{definition}[Path graph as an index]
Let $G = (V, E)$ be a graph, and let $G' = (V', E')$ be the \orderk{k} path graph of $G$ with key set $\mathcal{K}$. Pattern $X \in \Sigma^{\ast}$ \emph{matches} node $v \in V'$ if it prefix-matches string $G'.\gkey(v)$. If $V_{X}$ is the set of nodes $v \in V'$ matching pattern $X$, the set of \emph{occurrences} for the pattern is $G'.\gvalue(X) = G'.\gvalue(V_{X}) = \bigcup_{v \in V_{X}} G'.\gvalue(v)$.
\end{definition}

In the general case, queries using a path graph may result in false positives for patterns $X$ of length $\abs{X} > \abs{K}$, where $K$ is the shortest key in the key set $\mathcal{K}$. To avoid this, we restrict our attention to path graphs that are equivalent to de~Bruijn graphs.

\begin{definition}[Equivalent path graphs]
Let $G = (V, E)$ and $G' = (V', E')$ be two path graphs. We say that graphs $G$ and $G'$ are \emph{equivalent}, if we have $G.\gvalue(X) = G'.\gvalue(X)$ for all patterns $X \in \Sigma^{\ast}$.
\end{definition}

\begin{definition}[Pruned de~Bruijn graph]
Let $G = (V, G)$ be a graph, and let $G' = (V', E')$ be the \orderk{k} path graph of $G$ with key set $\mathcal{K}$. Path graph $G'$ is an \orderk{k} \emph{pruned de~Bruijn graph}, if it is equivalent to the \orderk{k} de~Bruijn graph of $G$, and $\abs{K} \le k$ for all $K \in \mathcal{K}$.
\end{definition}

Pruned de~Bruijn graphs can be built by merging nodes of a de~Bruijn graph. They are similar to manifold de Bruijn graphs \cite{Lin2014}.

\begin{lemma}[Pruning a de~Bruijn graph]\label{lemma:dbg-prune}
Let $G = (V, G)$ be a graph, let $G' = (V', E')$ be the \orderk{k} pruned de~Bruijn graph of $G$ with key set $\mathcal{K}$, and let $V_{K}$ be the set of nodes $v \in V'$ having string $K$ as a prefix of $G'.\gkey(v)$. If $\abs{V_{K}} > 0$ and $G'.\gvalue(u) = G'.\gvalue(v)$ for all $u, v \in V_{K}$, the path graph remains an \orderk{k} pruned de~Bruijn graph of $G$ after we remove $G'.\gkey(v)$ for all $v \in V_{K}$ from the key set and replace them with $K$.
\end{lemma}

\begin{proof}
Let $X \in \Sigma^{\ast}$ be a pattern, and let $V_{X} \subseteq V'$ be the set of nodes matching it. If the pattern does not prefix-match string $K$, we have $V_{X} \cap V_{K} = \emptyset$, and the changes in the key set do not affect the set of occurrences for pattern $X$.

Otherwise let $G'' = (V'', E'')$ be the path graph with adjusted key set, and let $v_{K} \in V''$ be the node corresponding to key $K$. Then
\begin{align*}
G''.\gvalue(X) & = G''.\gvalue(V_{X} \setminus V_{K}) \cup G''.\gvalue(v_{K}) \\
 & = G'.\gvalue(V_{X} \setminus V_{K}) \cup G'.\gvalue(V_{X} \cap V_{K}) = G'.\gvalue(X).
\end{align*}
\end{proof}

\begin{lemma}[Correctness]\label{lemma:dbl-correct}
Let $G = (V, G)$ be a graph, and let $G' = (V', E')$ be an \orderk{k} pruned de~Bruijn graph of $G$. For every pattern $X \in (\Sigma \setminus \set{\#, \$})^{\ast}$, let $V_{X}$ be the set of start nodes of paths $P$ in graph $G$ with $G.\glabel(P) = X$. Then $V_{X} \subseteq G'.\gvalue(X)$, and
$V_{X} = G'.\gvalue(X)$ if $\abs{X} \le k$.
\end{lemma}

% FIXME
% Introduce the de Bruijn graph
% Set of nodes of DBG matching the pattern -> set of kmers in S prefix-matching the pattern
% -> set of start nodes in G for paths matching the first min(k, |X|) characters of the pattern.
% If |X| > k, this may be a superset of start nodes matching the entire pattern.
% Pruned DBG equivalent to DBG -> holds with G' as well.
\begin{proof}
Foo
\end{proof}

\section{GCSA2}


\section{Construction}


\section{Experiments}


\section{Discussion}


% FIXME
\subparagraph*{Acknowledgements.}

I thank Richard Durbin, Erik Garrison, and Adam Novak for \dots


\bibliography{paper}


\end{document}
